### PR TITLE
Add rules.mk defaults for f103,f072,f042

### PR DESCRIPTION
**Pull request by zvecr**
## Description

As an attempt to resurrect parts of #5669, this PR aims to add defaults for a few of the more popular STM32 chips.

2nd pass will be to refactor boards to use these new defaults where possible
 
## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/mcu_selection.mk [hunk 1]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+  # Options to pass to dfu-util when flashing
+  DFU_ARGS ?= -d 0483:df11 -a 0 -s 0x08000000:leave
+  DFU_SUFFIX_ARGS ?= -p DF11 -v 0483
+endif
+
+ifneq ($(findstring STM32F072, $(MCU)),)
+  ## chip/board settings
+  # - the next two should match the directories in
+  #   <chibios>/os/hal/ports/$(MCU_FAMILY)/$(MCU_SERIES)
+  MCU_FAMILY ?= STM32
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  MCU_FAMILY = STM32
```

File: quantum/mcu_selection.mk [hunk 2]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+  MCU_SERIES ?= STM32F0xx
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  MCU_SERIES = STM32F0xx
```

File: quantum/mcu_selection.mk [hunk 3]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+
+  # Board: it should exist either in <chibios>/os/hal/boards/,
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= ST_STM32F072B_DISCOVERY
+
+  # Cortex version
+  MCU = cortex-m0
+
+  # ARM version, CORTEX-M0/M1 are 6, CORTEX-M3/M4/M7 are 7
+  ARMV ?= 6
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  ARMV = 6
```

File: quantum/mcu_selection.mk [hunk 4]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+  # Options to pass to dfu-util when flashing
+  DFU_ARGS ?= -d 0483:df11 -a 0 -s 0x08000000:leave
+  DFU_SUFFIX_ARGS ?= -p DF11 -v 0483
+endif
+
+ifneq ($(findstring STM32F042, $(MCU)),)
+  ## chip/board settings
+  # - the next two should match the directories in
+  #   <chibios>/os/hal/ports/$(MCU_FAMILY)/$(MCU_SERIES)
+  MCU_FAMILY ?= STM32
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  MCU_FAMILY = STM32
```

File: quantum/mcu_selection.mk [hunk 5]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+  MCU_SERIES ?= STM32F0xx
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  MCU_SERIES = STM32F0xx
```

File: quantum/mcu_selection.mk [hunk 6]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+
+  # Board: it should exist either in <chibios>/os/hal/boards/,
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= GENERIC_STM32_F042X6
+
+  # Cortex version
+  MCU = cortex-m0
+
+  # ARM version, CORTEX-M0/M1 are 6, CORTEX-M3/M4/M7 are 7
+  ARMV ?= 6
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  ARMV = 6
```

File: quantum/mcu_selection.mk [hunk 7]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+  # Options to pass to dfu-util when flashing
+  DFU_ARGS ?= -d 0483:df11 -a 0 -s 0x08000000:leave
+  DFU_SUFFIX_ARGS ?= -p DF11 -v 0483
+endif
+
+ifneq ($(findstring STM32F103, $(MCU)),)
+  ## chip/board settings
+  # - the next two should match the directories in
+  #   <chibios>/os/hal/ports/$(MCU_FAMILY)/$(MCU_SERIES)
+  MCU_FAMILY ?= STM32
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  MCU_FAMILY = STM32
```

File: quantum/mcu_selection.mk [hunk 8]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+  MCU_SERIES ?= STM32F1xx
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  MCU_SERIES = STM32F1xx
```

File: quantum/mcu_selection.mk [hunk 9]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
...
+
+  # Board: it should exist either in <chibios>/os/hal/boards/,
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= GENERIC_STM32_F103
+
+  # Cortex version
+  MCU = cortex-m3
+
+  # ARM version, CORTEX-M0/M1 are 6, CORTEX-M3/M4/M7 are 7
+  ARMV ?= 7
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  ARMV = 7
```

File: quantum/mcu_selection.mk [hunk 10]
```diff
@@ -23,17 +23,107 @@ ifneq ($(findstring STM32F303, $(MCU)),)
 
   # ARM version, CORTEX-M0/M1 are 6, CORTEX-M3/M4/M7 are 7
   ARMV ?= 7
```
Comment (fauxpark) [inline quantum/mcu_selection.mk]:
```suggestion
  ARMV = 7
```

**Comment by zvecr** [comment]
Seems like a good opportunity to align more stuff! Done in 00643367d31bb38f63c59457550da481136ecb55.

**Pull request merged**